keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,1190 @@
+adopt new format strings safely without breaking user workflows.
+
+#### K8s Update Widens CRD Validation
+
+In this case the user has a CRD with a field with following schema:
+
+```yaml
+type: string
+format: byte
+```
+
+In k8s 1.27 and below, this format disallows empty strings due to an implementation detail. However, in native types like `Secret`, byte is disallowed. Is Kubernetes had an update to CRDs to allow empty string for `byte`, this widens the validation for those CRDS which use it.
+
+This is fine when upgrading to new versions if Kubernetes: all existing resources still pass validation. But downgrading from an update that widens validation can be understood 
+semantically as an update that tightens validating.
+
+For example if they had unreletated issues with the new Kubernetes update and 
+wanted to downgrade:
+
+1. User updates to new Kubernetes version
+2. New CRs that were previously disallowed are allowed to be stored
+3. User is disappointed with K8s version and downgrades
+4. Controllers fail to update some resources, since they fail validation.
+
+If the CRD Validation Ratcheting feature were enabled the following occurs:
+
+1. User updates to new Kubernetes version
+2. New CRs that were previously disallowed are allowed to be stored
+3. User is disappointed with K8s version and downgrades
+4. Controllers are allowed to update their resources, as long as all their changes are valid.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+This KEP is scoped to CRDs, but if [KEP-3937](https://github.com/kubernetes/enhancements/pull/3938) is implemented then we should expect it to also affect native type validation.
+
+### Risks and Mitigations
+
+#### Detection of Breaking Schema Changes is Different
+
+Some users might have a process to detect breaking schema changes by applying
+their CRD and poking an existing resource. This KEP would make that process
+no longer reliable.
+
+##### Use CRD-Schema-Checker
+
+[CRD-Schema-Checker](https://github.com/openshift/crd-schema-checker) may be
+used to lint the CRD schema for breaking changes before applying it to a cluster.
+
+##### Use New Objects
+
+Alternatively, users may create new objects rather than updating old ones to 
+test for breaking schema changes. This feature has no effect on the creation 
+code path.
+
+
+#### Some CRD Validations Are Not Ratchetable
+
+Some usages of the OpenAPI schema are not possible to reliably ratchet without
+creating confusing exceptions, or rewriting completely the validation code used
+by Kubernetes. 
+
+##### Blacklisted Validations
+
+To mitigate this, validation failures in rules which have an
+ancestor of one of these types will not be ratcheted:
```

Comment:
My gut feeling is that to make this whole KEP feasible we should heavily restrict the OpenAPI validation constructs we can ratchet. Value validations that are basically leaves directly attached to structural schemas (the non-validation part in the sense of https://github.com/kubernetes/kubernetes/blob/d220ecb415485980ff02b7a010031bf279e9e067/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/structural.go#L34), i.e. without `any/all/oneOf`, `not` (the only `NestedValueValidation`).

The reason: you can replace the kube-openapi validation call on the root by multiple kube-openapi validation calls for each of those leaves.

This is a reasonable restriction because those nested value validations are not common, and they usually don't overlap (or we can assume if they do we can consider them to be atomic). Most of my example will go way, and we would not need to touch kube-openapi in the first place.

Reply:
Also wondering: if we collect the referenced paths in a value validation, we know what to compare for ratcheting logic. I.e. assign the list of referenced paths to every value validation. For 99% of them this is just `.` because they are leaves.

Reply:
Talking about https://github.com/kubernetes/kubernetes/blob/ecf54b8bd5fd7051a0931920a92793716ff82b90/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go#LL61C57-L61C57. Instead of having one giant validator for the whole schema, we can recursve over the structual schema and invoke kube-openapi validation multiple times with a subschema. That way we have full control over the field path and have access to the old object. And finally, the chance will be contained and easier to review.

Reply:
I'm implementing a basic prototype that attempts to follow the spirit of this simpler implementation idea here:

https://github.com/kubernetes/kube-openapi/pull/406

Normal forms/ slices not yet handled. But can get a basic idea of how objects are working.

Reply:
I have not attempted to implement exactly as you specified via structural schema since I started before reading that. I’ll try both and compare. But I think both approaches seem promising

Reply:
> My gut feeling is that to make this whole KEP feasible we should heavily restrict the OpenAPI validation constructs we can ratchet.

Agreed in the other thread. Updated KEP to restrict these constructs.